#!/usr/bin/env python

from __future__ import print_function

import rospy

import sys
import copy
import math
import moveit_commander

import moveit_msgs.msg
from moveit_msgs.msg import Constraints, JointConstraint, PositionConstraint, OrientationConstraint, BoundingVolume
from sensor_msgs.msg import JointState
from moveit_msgs.msg import RobotState
import geometry_msgs.msg
from geometry_msgs.msg import Quaternion, Pose, PoseStamped
from std_msgs.msg import String
from moveit_commander.conversions import pose_to_list
from baxter_moveit.msg import BaxterTrajectory
from baxter_moveit.srv import TrajectoryService, TrajectoryServiceRequest, TrajectoryServiceResponse

# Global variables
joint_names = ['left_s0', 'left_s1', 'left_e0', 'left_e1', 'left_w0', 'left_w1', 'left_w2']
trajectory_publisher = rospy.Publisher('baxter_joint_trajectory', BaxterTrajectory, queue_size=10)

# Between Melodic and Noetic, the return type of plan() changed. moveit_commander has no __version__ variable, so checking the python version as a proxy
if sys.version_info >= (3, 0):
    def planCompat(plan):
        return plan[1]
else:
    def planCompat(plan):
        return plan
       
"""
    Given the start angles of the robot, plan a trajectory that ends at the destination pose.
"""
def plan_trajectory(move_group, destination_pose, start_joint_angles): 
    current_joint_state = JointState()
    current_joint_state.name = joint_names
    current_joint_state.position = start_joint_angles

    moveit_robot_state = RobotState()
    moveit_robot_state.joint_state = current_joint_state
    move_group.set_start_state(moveit_robot_state)

    move_group.set_pose_target(destination_pose)
    move_group.set_goal_tolerance(0.00001)
    move_group.set_max_velocity_scaling_factor(0.8)
    
    plan = move_group.plan()

    if not plan:
        exception_str = """
            Trajectory could not be planned for a destination of {} with starting joint angles {}.
            Please make sure target and destination are reachable by the robot.
        """.format(destination_pose, destination_pose)
        exit(1)

    return planCompat(plan)

def plan_pick_and_place(req):
    response = TrajectoryServiceResponse()

    group_name = "left_arm"
    move_group = moveit_commander.MoveGroupCommander(group_name)

    # TODO: Make message type a list instead
    current_robot_joint_configuration = [
        math.radians(req.joints_input.joint_00),
        math.radians(req.joints_input.joint_01),
        math.radians(req.joints_input.joint_02),
        math.radians(req.joints_input.joint_03),
        math.radians(req.joints_input.joint_04),
        math.radians(req.joints_input.joint_05),
        math.radians(req.joints_input.joint_06),
    ]

    # Pre grasp - position gripper directly above target object
    pre_grasp_pose = plan_trajectory(move_group, req.pick_pose, current_robot_joint_configuration)

    previous_ending_joint_angles = pre_grasp_pose.joint_trajectory.points[-1].positions
    response.trajectories.append(pre_grasp_pose)

    # Grasp - lower gripper so that fingers are on either side of object
    pick_pose = copy.deepcopy(req.pick_pose)
    pick_pose.position.z -= 0.1
    grasp_pose = plan_trajectory(move_group, pick_pose, previous_ending_joint_angles)

    previous_ending_joint_angles = grasp_pose.joint_trajectory.points[-1].positions
    response.trajectories.append(grasp_pose)

    # Pick Up - raise gripper back to the pre grasp position
    pick_up_pose = plan_trajectory(move_group, req.pick_pose, previous_ending_joint_angles)

    previous_ending_joint_angles = pick_up_pose.joint_trajectory.points[-1].positions
    response.trajectories.append(pick_up_pose)
    
    # Waypoints following
    for i in range(len(req.waypoints)):
    	waypoint_pose = plan_trajectory(move_group, req.waypoints[i], previous_ending_joint_angles)

    	previous_ending_joint_angles = waypoint_pose.joint_trajectory.points[-1].positions
    	#del waypoint_pose.joint_trajectory.points[-1]
    	response.trajectories.append(waypoint_pose)

    # Place - move gripper to desired placement position
    place_pose = plan_trajectory(move_group, req.place_pose, previous_ending_joint_angles)

    previous_ending_joint_angles = place_pose.joint_trajectory.points[-1].positions
    response.trajectories.append(place_pose)

    move_group.clear_pose_targets()
    
    jointsMsg = BaxterTrajectory()
    jointsMsg.trajectory = response.trajectories
    trajectory_publisher.publish(jointsMsg)

    return response


def moveit_server():
    moveit_commander.roscpp_initialize(sys.argv)
    rospy.init_node('baxter_moveit_server')
    
    s = rospy.Service('baxter_moveit_trajectory', TrajectoryService, plan_pick_and_place)
        
    scene = moveit_commander.PlanningSceneInterface()
    # robot = moveit_commander.RobotCommander()

    rospy.sleep(1)

    # Add collision objects to the scene
    p = PoseStamped()
    p.header.frame_id = "world"
    p.pose.position.x = 0
    p.pose.position.y = 0.
    p.pose.position.z = 0.4
    scene.add_box("table", p, (2, 2, 0.67))
    
    print("Ready to plan")
    rospy.spin()


if __name__ == "__main__":
    moveit_server()
